pkg/scheduler/apis/config/v1beta1/defaults.go

```diff
@@ -200,7 +200,7 @@ func SetDefaults_PodTopologySpreadArgs(obj *v1beta1.PodTopologySpreadArgs) {
 		// SelectorSpread plugin.
 		return
 	}
-	if obj.DefaultConstraints == nil {
+	if len(obj.DefaultConstraints) == 0 && !obj.DisableDefaultConstraints {
```

Comment:
I guess people were providing an empty list to avoid getting the default? Yeah, that's bad.

This conditional defaulting is going to give you other problems though.

I actually recommend deleting this defaulting completely and making the consumer of the config do this logic.

In particular I think you should avoid mixing system defaults and user provided defaults in a single field in this layer.

Reply:
> I guess people were providing an empty list to avoid getting the default? Yeah, that's bad.

Yeah, my bad. OTOH, the system defaulting is an alpha feature.

> I actually recommend deleting this defaulting completely and making the consumer of the config do this logic.

Even though this code is within pkg/scheduler, i.e. the consumer?

> This conditional defaulting is going to give you other problems though.

Could you explain?

Also, having this in defaulting makes it easier to share what configuration is the plugin actually using, once we resolve #93718

Reply:
> Even though this code is within pkg/scheduler, i.e. the consumer?

The defaulting function is called during deserialization, which means original user intent is gone by the time the eventual actual consumer gets the go structures.

> > This conditional defaulting is going to give you other problems though.

> Could you explain?

We're adding support for static defaulting, but it won't be able to replace conditional defaulting which depends on multiple fields. I don't know if I can make a great argument for this off the top of my head, but we've found that complicated (e.g. conditional or multi-field) defaulting logic is almost always regretted later.

It's less important because this is a config file, but I'd still try to avoid it.

In my mind there's two concepts that are mixed here, "what value does this field have if it's not specified", and "what does the application do given certain input". This distinction isn't imposed on us by the laws of the universe, but imposing it on ourselves makes the codebase easier to understand, e.g. less spooky action at a distance.

If the system were not ever going to evolve then I'd be fine with the first version of this I saw. But given that it will evolve, I think it's easier to evolve this logic when it's in the application layer and not the defaulting (deserialization) layer.

I hope that makes sense, sorry :)

> easier to share what configuration is the plugin actually using

I'm not sure how? It sounds like stuff that'd go in a "status" section except we don't have those for config files.